### PR TITLE
Update cryptography to 42.0.4 and update certdir (SYN-3552, SYN-6860)

### PR DESCRIPTION
**Pull request by vEpiphyte**
- Cryptography update addresses older version of cryptography package containing CVE-2023-50782 & CVE-2024-26130
- certdir now uses cryptography X509 objects and RSA private key objects, instead of PyOpenSSL X509 and Pkey objects. This is largely due to the removal of APIs from PyOpenSSL which we were utilizing for PKCS12 support and the guidance from PyOpenSSL project to not utilize the ``Crypto`` module in new projects as it is considered deprecated in favor of Cryptography. Per prior discussion, there should be no API stability concerns related to this change since the CertDir class is not exposed via telepath or storm apis.
- certdir is now fully typed. This identified issues where we were declaring bytes as inputs on certdir and Cortex was passing in PEM strings instead of bytes.
- Remove PyOpenSSL use where it is possible to do so. We now only use it for doing X509 path building and certificate verification, eventually we'll be able to remove this in favor of APIs provided by Cryptography (  see https://github.com/pyca/cryptography/issues/10393 https://github.com/pyca/cryptography/issues/10034 )

### CODE REVIEW

File: synapse/lib/certdir.py [hunk 1]
```diff
@@ -193,288 +229,356 @@ def __init__(self, path=None):
...
         Generates a CA keypair.
 
         Args:
-            name (str): The name of the CA keypair.
-            signas (str): The CA keypair to sign the new CA with.
-            outp (synapse.lib.output.Output): The output buffer.
+            name: The name of the CA keypair.
+            signas: The CA keypair to sign the new CA with.
+            outp: The output buffer.
+            save: Safe the certificate and key to disk.
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:258]:
```suggestion
            save: Save the certificate and key to disk.
```

File: synapse/lib/certdir.py [hunk 2]
```diff
@@ -193,288 +229,356 @@ def __init__(self, path=None):
...
         fullpath = s_common.genpath(*path)
         self.pathrefs[fullpath] -= 1
         if self.pathrefs[fullpath] <= 0:
             self.certdirs.remove(fullpath)
             self.pathrefs.pop(fullpath, None)
 
-    def genCaCert(self, name, signas=None, outp=None, save=True):
+    def genCaCert(self, name: str,
+                  signas: StrOrNone = None,
+                  outp: OutPutOrNone =None,
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:249]: Should these be `= None` or `=None`? Seems to be a mix of spacing in use

File: synapse/lib/certdir.py [hunk 3]
```diff
@@ -524,92 +630,100 @@ def genClientCert(self, name, outp=None):
...
-    def valUserCert(self, byts, cacerts=None):
+    def valUserCert(self, byts: bytes, cacerts: Union[List[c_x509.Certificate] | None] =None) -> c_x509.Certificate:
         '''
         Validate the PEM encoded x509 user certificate bytes and return it.
 
         Args:
-            byts (bytes): The bytes for the User Certificate.
-            cacerts (tuple): A tuple of OpenSSL.crypto.X509 CA Certificates.
+            byts: The bytes for the User Certificate.
+            cacerts: A tuple of CA Certificates to use for validating the user cert..
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:648]:
```suggestion
            cacerts: A tuple of CA Certificates to use for validating the user cert.
```

File: synapse/lib/certdir.py [hunk 4]
```diff
@@ -629,69 +743,69 @@ def getCaCerts(self):
...
 
         Examples:
-            Get the path to the CA certificate for the CA "myca":
+            Get the path to the CA certificate for the CA "myca"::
 
                 mypath = cdir.getCACertPath('myca')
 
         Returns:
-            str: The path if exists.
+            The path if exists.
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:759]:
```suggestion
            The path, if exists.
```

File: synapse/lib/certdir.py [hunk 5]
```diff
@@ -629,69 +743,69 @@ def getCaCerts(self):
...
+            name: The name of the CA keypair.
 
         Examples:
             Get the path to the private key for the CA "myca":
 
                 mypath = cdir.getCAKeyPath('myca')
 
         Returns:
-            str: The path if exists.
+            The path if exists.
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:796]:
```suggestion
            The path, if exists.
```

File: synapse/lib/certdir.py [hunk 6]
```diff
@@ -702,120 +816,121 @@ def getClientCert(self, name):
...
+            name: The name of the client keypair.
 
         Examples:
             Get the path to the client certificate for "myuser":
 
                 mypath = cdir.getClientCertPath('myuser')
 
         Returns:
-            str: The path if exists.
+            The path if exists.
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:836]:
```suggestion
            The path, if exists.
```

File: synapse/lib/certdir.py [hunk 7]
```diff
@@ -702,120 +816,121 @@ def getClientCert(self, name):
...
         Examples:
-            Get the path to the CA cert which issue the cert for "myhost":
+
+            Get the path to the CA cert which issue the cert for "myhost"::
 
                 mypath = cdir.getHostCaPath('myhost')
 
         Returns:
-            str: The path if exists.
+            The path if exists.
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:857]:
```suggestion
            The path, if exists.
```

File: synapse/lib/certdir.py [hunk 8]
```diff
@@ -702,120 +816,121 @@ def getClientCert(self, name):
...
+            name: The name of the host keypair.
 
         Examples:
             Get the path to the host certificate for the host "myhost":
 
                 mypath = cdir.getHostCertPath('myhost')
 
         Returns:
-            str: The path if exists.
+            The path if exists.
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:901]:
```suggestion
            The path, if exists.
```

File: synapse/lib/certdir.py [hunk 9]
```diff
@@ -827,74 +942,74 @@ def getHostKeyPath(self, name):
...
 
         Examples:
-            Get the path for the user cert for "myuser":
+            Get the path for the user cert for "myuser"::
 
                 mypath = cdir.getUserCertPath('myuser')
 
         Returns:
-            str: The path if exists.
+            The path if exists.
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:996]:
```suggestion
            The path, if exists.
```

File: synapse/lib/certdir.py [hunk 10]
```diff
@@ -906,65 +1021,65 @@ def getUserForHost(self, user, host):
...
 
         Examples:
-            Get the path to the user key for "myuser":
+            Get the path to the user key for "myuser"::
 
                 mypath = cdir.getUserKeyPath('myuser')
 
         Returns:
-            str: The path if exists.
+            The path if exists.
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:1055]:
```suggestion
            The path, if exists.
```

File: synapse/lib/certdir.py [hunk 11]
```diff
@@ -193,288 +229,356 @@ def __init__(self, path=None):
...
 
         Examples:
-            Make a host keypair named "myhost":
+            Make a host keypair named "myhost"::
 
                 myhostkey, myhostcert = cdir.genHostCert('myhost')
 
         Returns:
-            ((OpenSSL.crypto.PKey, OpenSSL.crypto.X509)): Tuple containing the private key and certificate objects.
+            Tuple containing the private key and certificate objects. Private key may be none when signing a CSR.
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:313]:
```suggestion
            Tuple containing the private key and certificate objects. Private key may be None when signing a CSR.
```

File: synapse/tests/test_lib_cell.py [hunk 12]
```diff
@@ -12,6 +12,8 @@
 
 from unittest import mock
 
+import cryptography.x509 as c_x509
+
```
Comment (Cisphyx) [inline synapse/tests/test_lib_cell.py:16]: looks like we don't need to `from OpenSSL import crypto` here anymore

File: synapse/lib/certdir.py (hunk 2 above)
Comment (vEpiphyte) [reply synapse/lib/certdir.py]: `= None` is preferred.

File: synapse/lib/certdir.py [hunk 13]
```diff
@@ -101,41 +135,43 @@ def revoke(self, cert):
...
 
     def _verify(self, cert):
         # Verify the cert was signed by the CA in self.name
         cacert = self.certdir.getCaCert(self.name)
         store = crypto.X509Store()
-        store.add_cert(cacert)
+        store.add_cert(crypto.X509.from_cryptography(cacert))
         store.set_flags(crypto.X509StoreFlags.PARTIAL_CHAIN)
-        ctx = crypto.X509StoreContext(store, cert,)
+        ctx = crypto.X509StoreContext(store, crypto.X509.from_cryptography(cert), )
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:155]: Is the trailing comma actually required here?

Comment (Cisphyx) [inline synapse/lib/certdir.py:155]:
```suggestion
        ctx = crypto.X509StoreContext(store, crypto.X509.from_cryptography(cert))
```

File: synapse/lib/certdir.py [hunk 14]
```diff
@@ -621,77 +737,77 @@ def getCaCerts(self):
...
+    def getCaKeyPath(self, name: str) -> StrOrNone:
         '''
         Gets the path to a CA key.
 
         Args:
-            name (str): The name of the CA keypair.
+            name: The name of the CA keypair.
 
         Examples:
             Get the path to the private key for the CA "myca":
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:793]:
```suggestion
            Get the path to the private key for the CA "myca"::
```

File: synapse/lib/certdir.py [hunk 15]
```diff
@@ -621,77 +737,77 @@ def getCaCerts(self):
...
+    def getClientCert(self, name: str) -> Pkcs12OrNone:
         '''
         Loads the PKCS12 archive object for a given user keypair.
 
         Args:
-            name (str): The name of the user keypair.
+            name: The name of the user keypair.
 
         Examples:
             Get the PKCS12 object for the user "myuser":
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:813]:
```suggestion
            Get the PKCS12 object for the user "myuser"::
```

File: synapse/lib/certdir.py [hunk 16]
```diff
@@ -702,120 +818,121 @@ def getClientCert(self, name):
...
+    def getClientCertPath(self, name: str) -> StrOrNone:
         '''
         Gets the path to a client certificate.
 
         Args:
-            name (str): The name of the client keypair.
+            name: The name of the client keypair.
 
         Examples:
             Get the path to the client certificate for "myuser":
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:833]:
```suggestion
            Get the path to the client certificate for "myuser"::
```

File: synapse/lib/certdir.py [hunk 17]
```diff
@@ -702,120 +818,121 @@ def getClientCert(self, name):
...
+    def getHostCertPath(self, name: str) -> StrOrNone:
         '''
         Gets the path to a host certificate.
 
         Args:
-            name (str): The name of the host keypair.
+            name: The name of the host keypair.
 
         Examples:
             Get the path to the host certificate for the host "myhost":
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:898]:
```suggestion
            Get the path to the host certificate for the host "myhost"::
```

File: synapse/lib/certdir.py [hunk 18]
```diff
@@ -1364,72 +1538,70 @@ def _genPkeyCsr(self, name, mode, outp=None):
...
-    def _getPathBytes(self, path):
+    def _getPathBytes(self, path: str) -> BytesOrNone:
         if path is None:
             return None
         return s_common.getbytes(path)
 
-    def _getPathJoin(self, *paths):
+    #
+    def _getPathJoin(self, *paths: str) -> str:
+        '''Get the base certidr path + paths'''
```
Comment (Cisphyx) [inline synapse/lib/certdir.py:1552]:
```suggestion
        '''Get the base certdir path + paths'''
```

**Review by MichaelSquires — CHANGES_REQUESTED** [review]

File: synapse/lib/certdir.py [hunk 19]
```diff
@@ -5,23 +5,54 @@
...
+CertOrNone = Union[c_x509.Certificate | None]
+PkeyOrNone = Union[c_rsa.RSAPrivateKey | c_dsa.DSAPrivateKey | None]
+PkeyAndCert = Tuple[c_rsa.RSAPrivateKey, c_x509.Certificate]
+PkeyAndBuilder = Tuple[c_rsa.RSAPrivateKey, c_x509.CertificateBuilder]
+PrivKeyPubKeyBuilder = Tuple[Union[c_rsa.RSAPrivateKey | None], c_types.PublicKeyTypes, c_x509.CertificateBuilder]
+Pkey = Union[c_rsa.RSAPrivateKey | c_dsa.DSAPrivateKey]
+Pkcs12OrNone = Union[c_pkcs12.PKCS12KeyAndCertificates | None]
+PkeyOrNoneAndCert = Tuple[Union[c_rsa.RSAPrivateKey | None], c_x509.Certificate]
+# Used for handling CSRs
+PubKeyOrNone = Union[c_types.PublicKeyTypes | None]
```
[the remark this answers is not in the document]
Comment (vEpiphyte) [reply synapse/lib/certdir.py:56]: Kind of went back and forth here - we several of these are also used as return value types too, which is more generic than `Optional`.

**Review by Cisphyx — APPROVED** [review]

**Review by mikemoritz — APPROVED** [review]

**Pull request merged**